### PR TITLE
Handle KeyboardEvent without virtual key code

### LINKED ISSUE #12616: Can't enter <äöüß#+ keys on german keyboard
Follow up of #12189. Tested on commit b41bf4f1f30b8f4eec92909c8971bc02d5422b50 .

Trying to press one of the `<äöüß#+` keys on a german keyboard yields in no key entered in input fields. Note that #, <, and + are even within the ascii character set. Other keys like z work properly (if it were a qwerty layout, it would register a y, but it correctly registers a z). `,.-` work as well.

Image of the german keyboard layout: https://en.wikipedia.org/wiki/File:KB_Germany.svg

### LINKED ISSUE #15240: Handle KeyEvent with no virtual key code.
When a KeyboardInput has no [virtual key code](http://doc.servo.org/glutin/enum.VirtualKeyCode.html), servo does [nothing](https://github.com/servo/servo/blob/master/ports/glutin/window.rs#L396). It may be possible to have some characters such as ```é```, ```µ``` which doesn't have any virtualkeycode but we should also send them to the compositor.

### PR DESCRIPTION
**Pull request by charlesvdv**
This PR allows to properly handling KeyboardEvent without a virtual key code. 

These PR is not completed *yet* but I prefer start a discussion to see if my approach is correct to you. (For exemple I have to remove the [```Result``` over here](https://github.com/servo/servo/compare/master...charlesvdv:fix-keyboard?expand=1#diff-f3afdd0d8092731ed5c423ce2e8dfd70R620)). 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #15240 (github issue number if applicable).
- [ ] Theses changes *should* fix #12616

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15800)

### CODE REVIEW

File: ports/glutin/window.rs [hunk 1]
```diff
@@ -345,16 +345,16 @@ impl Window {
...
-                    VirtualKeyCode::LWin => self.toggle_modifier(LEFT_SUPER),
-                    VirtualKeyCode::RWin => self.toggle_modifier(RIGHT_SUPER),
+                    Some(VirtualKeyCode::LControl) => self.toggle_modifier(LEFT_CONTROL),
+                    Some(VirtualKeyCode::RControl) => self.toggle_modifier(RIGHT_CONTROL),
+                    Some(VirtualKeyCode::LShift) => self.toggle_modifier(LEFT_SHIFT),
+                    Some(VirtualKeyCode::RShift) => self.toggle_modifier(RIGHT_SHIFT),
+                    Some(VirtualKeyCode::LAlt) => self.toggle_modifier(LEFT_ALT),
+                    Some(VirtualKeyCode::RAlt) => self.toggle_modifier(RIGHT_ALT),
+                    Some(VirtualKeyCode::LWin) => self.toggle_modifier(LEFT_SUPER),
+                    Some(VirtualKeyCode::RWin) => self.toggle_modifier(RIGHT_SUPER),
```
Comment (KiChjang) [inline ports/glutin/window.rs]: Why is this an improvement?

Comment (charlesvdv) [inline ports/glutin/window.rs]: [This is the improvements](https://github.com/servo/servo/pull/15800/files#diff-f3afdd0d8092731ed5c423ce2e8dfd70L396) because now we are handling correctly the case where virtual key code is ```None```.

File: ports/glutin/window.rs [hunk 2]
```diff
@@ -345,16 +345,16 @@ impl Window {
                     self.pending_key_event_char.set(Some(ch));
                 }
             }
-            Event::KeyboardInput(element_state, scan_code, Some(virtual_key_code)) => {
```
Comment (KiChjang) [inline ports/glutin/window.rs]: Removing the match arm you linked should just be fine, why did you need to remove this `Some()`?

Comment (KiChjang) [inline ports/glutin/window.rs]: Oh I see, looks like you're rolling in functionality for handling Some and None into the same match arm. Would it not be ok to just make a separate match arm instead specifically for the None case?

Comment (charlesvdv) [inline ports/glutin/window.rs]: Because before we were just handling the case with Some(virtual_key) but now we handle the case None (which doesn't change much the code so I put both on the same case).

File: ports/glutin/window.rs [hunk 3]
```diff
@@ -384,17 +384,17 @@ impl Window {
...
-                        ElementState::Released => KeyState::Released,
-                    };
-                    let modifiers = Window::glutin_mods_to_script_mods(self.key_modifiers.get());
-                    self.event_queue.borrow_mut().push(WindowEvent::KeyEvent(ch, key, state, modifiers));
-                }
-            }
-            Event::KeyboardInput(_, _, None) => {
-                debug!("Keyboard input without virtual key.");
+                let key = match virtual_key_code {
+                    Some(v) => Window::glutin_key_to_script_key(v).unwrap(),
```
Comment (KiChjang) [inline ports/glutin/window.rs]: Why is it OK to assume the unwrap wouldn't panic?

Comment (charlesvdv) [inline ports/glutin/window.rs]: Because I will remove the ```Result``` from [here](https://github.com/servo/servo/pull/15800/files#diff-f3afdd0d8092731ed5c423ce2e8dfd70R620) but before doing so I prefer to see if you are ok with the changes 😄

File: ports/glutin/window.rs (hunk 2 above)
Comment (KiChjang) [inline ports/glutin/window.rs]: I'm actually more of the opinion that we should split this up into different match arms. The common code between the two match arms should just be refactored into a separate function, since this arm is already pretty big as it is. What do you think?

Comment (charlesvdv) [inline ports/glutin/window.rs]: I'm totally ok with that!

File: ports/glutin/window.rs (hunk 3 above)
Comment (KiChjang) [inline ports/glutin/window.rs]: Can't seem to follow the link you posted.

Comment (charlesvdv) [inline ports/glutin/window.rs]: Oh sorry. [This one should works!](https://github.com/servo/servo/blob/master/ports/glutin/window.rs#L620)

**Comment by KiChjang** [comment]
Changes look good to me, but I see that @jdm has raised some questions in the relevant issue. Handing this over to him for approval.
r? @jdm

**Comment by jdm** [comment]
I'm not available for reviewing code right now, so I'm going to hand this off to @Ms2ger.
r? @Ms2ger

**Review by Ms2ger — COMMENTED** [review]
I found it pretty hard to review, so I tried to split up your the refactoring from the actual changes: <https://github.com/servo/servo/tree/keyboard-virtual>.

Would you mind pushing that branch to your PR? That should help to evaluate it.

**Comment by nox** [comment]
That changes from your last fixup commit needs to be properly squashed into the commits where they belong.

**Comment by nox** [comment]
You don't need `Key::Unidentified`. Instead keep returning `Result<Key, ()>`, and use an `Option<Key>` with `None` where you used `Key::Unidentified`.

**Comment by nox** [comment]
@jonathandturner Think you could take a look at this? It became hard to rebase thanks to #16198.

**Comment by jonathandturner** [comment]
@nox - apologies.  Wish I could help, though unfortunately too busy/out-of-the-loop to help fix.

**Comment by dralley** [comment]
I think this PR (as well as possibly the mentioned issue) are obsolete, it seems like Servo's keyboard event handling was made more spec-compliant by https://github.com/servo/servo/commit/0ccaa7e1a9e9bf9472d869576019b9cda350ad87#diff-8dc9268416cd4f037d50206e76c73cf9

The linked issue also talks about Glutin limitations, I'm not sure if that is still relevant anymore either.

**Comment by jdm** [comment]
It certainly could still be relevant, since glutin just exported a bunch of input-related types and APIs from winit. We still rely on winit, so we probably still have the same input-related problems. However, this PR is not receiving any attention so I'm going to go ahead and close it.
